### ISSUE TITLE
Problems compiling FP IDE...

### ISSUE THREAD

**Issue by funlw65** [fredvs/freepascal-ootb#1]
Hi Fred,
I see that in the release there is no FP IDE present, is the  only IDE released with FreePascal and it should not be left aside. Also, if you try to compile FP with ootb, it fails...

**Comment by rchastain** [comment]
Hello! Please could you say how you (try to) compile FPIDE, and which source you use, so we can try too?

**Comment by funlw65** [comment]
ootb is on a global PATH, and I try to compile fp.pas from inside it's folder on the 3.2.2 sources, issuing a command line from terminal (linux here).

**Comment by funlw65** [comment]
Yeah, applications that use FV, if successfully compiled, does not work. And not only ootb has this problem, but also a patched (to link to the oldest C library) standard FPC using fpcupdeluxe gives the same result - apps that are using FV are not working.  

Happened on LinuxMiint LMDE 6

Once the patch is removed and a standard FPC is used, FV is usable again.

**Comment by fredvs** [comment]
Hello Funlw65.

May I ask you what is "FV" and what patch is "patch is removed "?
I will jump to the compilation of FPIDE asap and see what is wrong.

**Comment by fredvs** [comment]
Hello Funlw65.

>Also, if you try to compile FP with ootb, it fails...

I have compiled FP IDE from source fpc-ootb using Lazarus and setting fpc-ootb-64 as compiler on Xubuntu 22.04 + Xfce 64 bit.

![laz_fpcootb](https://github.com/fredvs/freepascal-ootb/assets/3421249/cd306d34-aeb4-4fd6-9040-2ffd3470ab8f)

Using **/freepascal-ootb-main/packages/ide/fpx64.lpi** project file and Lazarus.
It compiles OK.

![fpide_compile](https://github.com/fredvs/freepascal-ootb/assets/3421249/bcb7207b-6da1-4c19-8b8d-d83d63a97589)

And here the result of running:

![fpide_fpc-ootb](https://github.com/fredvs/freepascal-ootb/assets/3421249/c69c5902-2be3-4453-a652-937ac83c2f7b)

Here all seems ok. 

The zip file of fp ide binary for Linux 64 bit:  [fp64.zip](https://github.com/fredvs/freepascal-ootb/files/15204309/fp64.zip)

The zip file of fp ide binary for Linux 32 bit: [fp32.zip](https://github.com/fredvs/freepascal-ootb/files/15204291/fp32.zip)

Note that I did not find a working script to compile it without using the **fpx64.lpi** project and Lazarus.
Using, like explained in readme.txt, **make gdb** did not work.

Maybe the binary **fp** should be added in the release in /fpc_ootb/tools/.
[EDIT]  Just now: **fp** was added in /fpc_ootb/tools/ in the zip files of the release https://github.com/fredvs/freepascal-ootb/releases/tag/3.2.2 .

Do you see other binaries that should be added?

-----------------------------------

>Once the patch is removed and a standard FPC is used, FV is usable again.

What patch are you referring?

---------------------------

[EDIT] With this command line on terminal, the compilation is ok here:

First go in the directory of freepascal-ootb source:

`cd /directory-of/freepascal-ootb-main`

Then paste that command line adapting **_/directory-of/fpc-ootb-xx_** with your path of fpc-ootb compiler.
If you want that all compiled objects are in a separate directory, change **-FUpackages/ide/** with  your output directory wanted **-FU/your-directory-of-output/**.

For Linux 64 bit:
`/directory-of/fpc-ootb-64 -Mfpc -Scgi -CX -Cg -Os1 -Xs -XX -l -vewnhibq -Fipackages/ide -Ficompiler -Ficompiler/x86_64 -Fucompiler -Fucompiler/x86 -Fucompiler/x86_64 -Fucompiler/systems -Fupackages/rtl-extra/src/inc -Fupackages/fv/src -Fupackages/ide/ -FUpackages/ide/ -FEpackages/ide/ -opackages/ide/fp -dx86_64 -dnodebug -B packages/ide/fp.pas
`

For Linux 32 bit:
`/directory-of/fpc-ootb-32 -Mfpc -Scgi -CX -Cg -Os1 -Xs -XX -l -vewnhibq -Fipackages/ide -Ficompiler -Ficompiler/i386 -Fucompiler -Fucompiler/x86 -Fucompiler/i386 -Fucompiler/systems -Fupackages/rtl-extra/src/inc -Fupackages/fv/src -Fupackages/ide/ -FUpackages/ide/ -FEpackages/ide/ -opackages/ide/fp -di386 -dnodebug -B packages/ide/fp.pas
`

Fre;D

**Comment by funlw65** [comment]
Thank you Fred, I will try it. I'm referring to the patch that is added to fpc sources to be able to produce an executable linked to the oldest glibc possible to run on older Linux stations.

If I prepare such fpc compiler using the fpcupdeluxe, then compile the example that comes with FV, it won't work. If I use a standard fpc instead, the example will work just fine.

I'm glad yours is working just fine. And thank you for the inclusion of FP IDE. I have no knowledge about other tools that might be included.

**Comment by fredvs** [comment]
> I'm referring to the patch that is added to fpc sources to be able to produce an executable linked to the oldest glibc possible to run on older Linux stations.

Ha, fpc has done a patch finally, this is great.

By the way, **FP** IDE himself does not link to GLIBC:

`> readelf --dyn-syms -W fp | grep GLIBC`

result:
 ` >    `

**Comment by funlw65** [comment]
Sorry fred,  the patch is made by the fpcupdeluxe utility, not by the developers of the fpc. Sorry for the confusion.

I successfully compiled the FV example testapp.pas, using your command line, now I have to figure out how to find the configuration equivalent from inside the FP IDE to successfully compile the same example.

**Comment by fredvs** [comment]
>now I have to figure out how to find the configuration equivalent from inside the FP IDE to successfully compile the same example

Imho, fp has his own compiler and if you have build it using fpc-ootb, it should have the same fixes.
See here:
https://forum.lazarus.freepascal.org/index.php/topic,37054.msg247818.html#msg247818

fp uses fp.cfg for initializing his compiler and maybe you have to define de -Fu directories of fpc-ootb.

**Comment by funlw65** [comment]
It is easier to get all the units in a single path, using the wildcard.. I missed all the other parameters but then that was an easy fix, FP has a line input where you add all the extra flags you need, so it worked soon after I complained... Thank you for the help. I prefer your "distro" of fpc for Linux. Actually, I prefer having it in my home folder instead of spreading it all over /usr... Linux is a multiuser operating system, but this laptop is mine, if you get what I mean :D

**Comment by funlw65** [comment]
> > now I have to figure out how to find the configuration equivalent from inside the FP IDE to successfully compile the same example
> 
> Imho, fp has his own compiler and if you have build it using fpc-ootb, it should have the same fixes. 

Sooo, that means that fp binary is all I need if I want? In some way that is awesome! Good to know!

Edit: Yep! That is awesome, just tested it! That means I can have two separate packages. Yours for more complex, crosscompiling stuff, and the other with only FP and your units. Both packages very small! Awesome!

**Comment by fredvs** [comment]
Hello Funlw65

I have done some modifs in **fpide** source to make it out-of-the-box like **fpc-ootb**.
You may now use the macro **$FPCBINDIR** for the fpc units in the **fp.cfg**
No need to use the full path.
With **fpide** it is done with menu Options/Directories.

![fpide_dir](https://github.com/fredvs/freepascal-ootb/assets/3421249/71d38c5e-91aa-4504-b1cb-ba82713e4d4d)

The **fp.cfg** is created in the working directory when you start **fpide** and will be used for compilation of the file loaded (as far it is what I understood).
The default **fp.cfg** used when creating config files ("Copy existing") is located in /tools/text.

![fpide_cfg](https://github.com/fredvs/freepascal-ootb/assets/3421249/e8b8c671-10ab-4508-8869-d4deee2b4d1e)

Note that "**fp**" binary was renamed into "**fpide-ootb**" and placed alongside with **fpc-ootb** (because of **$FPCBINDIR**).

The zip in release are updated.
https://github.com/fredvs/freepascal-ootb/releases/tag/3.2.2

Fre;D

**Comment by funlw65** [comment]
You know what is weird? I used $FPCBINDIR in fp.cfg without downloading your modified fp and it works out of the box!?  Even before reading this last post of yours... 

It might be possible that I downloaded a version that had the modifications but not the change in the executable name?

**Comment by fredvs** [comment]
Yes but when you load fp-ide from a directory without fp.cfg included, it will create a fp.cfg without the -Fudirectories (or from official fpc).

With the last patch, the default fp.cfg has all that $FPCBINDIR already done.

Anyway, I did not test when fpide is inside /tools, if it works of course it is better than alongside with fpc-ootb.

**Comment by fredvs** [comment]
I just tested when fpide is in the directory /tools but here compilation does not work with $FPCBINDIR in fp.cfg.
Here it works only if fpide is in same directory as fpc-ootb.

>It might be possible that I downloaded a version that had the modifications but not the change in the executable name?

Is the name of fp renamed into fpide-ootb and placed in the same directory as fpc-ootb?
If yes, it is indeed the last zip.

**Comment by funlw65** [comment]
No, is not the new one, is named fp and it was in tools subfolder.

But now I have it in its own folder,  something like this:
$HOME/fp/ 
$HOME/fp/units
$HOME/fp/libs
$HOME/fp/inc
$HOME/fp/projects

In $HOME/fp I have the fp binary. fpc.cfg has that $FPCBINDIR and it works. But I have your ootb in  the folder named fpc and a symlink to your binary, named fpc... and it is in PATH. It might be possible that it launches fpc?

Edit: Nope, I renamed fpc folder and still works!

**Comment by funlw65** [comment]
The fp binary that I have (compiled by you - is from the previous package) has 4895864 bytes in size.

**Comment by funlw65** [comment]
Maybe is the way I copied the units?

Those are directly in the units folder, and there is no x86_64-linux intermediary folder.

By example, the a52 unit (it's own folder) is in:
  units/a52
instead of:
  units/x86_64-linux/a52

I mean, as I understand, the embedded pascal compiler can serve only one target so, in my mind, x86_64-linux had no meaning anymore. So I got rid of it. 

May be that the FP IDE mimics the behavior of the old Turbo Pascal? An "old unit searching scheme"? In that case, $FPCBIN DIR is ignored and the units are found in that way...

Edit: Nope, it does not compile without $FPCBINDIR ...  I have no other ideas...  BTW Fred, FP has inside its own compiler, or it includes the one that comes with the distro? Because that is the last posibility in solving "the case" :P . If is true, it had already the required modifications.

**Comment by fredvs** [comment]
Ha, I did not try this: copy fp.cfg into the same directory as FP IDE, maybe it takes this one as "default" and so it is the reason why it works for you.  I will try it asap.

Could you try this, when FP IDE is loaded, click on menu "Debug" / "Usr screen" and see what directory of fp.cfg is used?


![fpide_cfg2](https://github.com/fredvs/freepascal-ootb/assets/3421249/099e1296-178d-49a9-af30-b05fe6a0d59c)


> BTW Fred, FP has inside its own compiler, or it includes the one that comes with the distro? 

The FP has inside his own compiler compiled using the sourcefpc-ootb.
So mainly the compiler inside FP is the same as the compiler fpc-ootb.

**Comment by funlw65** [comment]
It uses the fp.cfg from it's main folder.  /home/user/fp 

user@shuttle:~/fp$ ./fp
* Free Pascal IDE Version 1.0.12 [2024/05/03]
* Compiler Version 3.2.2_glibc225
* Using configuration files from: /home/user/fp/

**Comment by fredvs** [comment]
OK so for example in console do:

`> cd /directory/of/your/project`

Then load fp:

`> /directory/of_fp/fp`

Normally, if there is  no fp.cfg in /directory/of/your/project there should be a message about missing config file and proposition to create one.

I will test asap when fp.cfg is inside same directory as fp but, it seems to me that without the last modif the new fp.cfg created will have directory of official fpc.

But maybe I miss something.

**Comment by funlw65** [comment]
Ok to put you into context, this is the tree of my fp installation:

![Screenshot from 2024-05-06 14-26-16](https://github.com/fredvs/freepascal-ootb/assets/804555/2468c9e2-c46b-4d58-9db2-92c6fb5a9ed6)

And in my main folder is only fp:
![Screenshot from 2024-05-06 14-26-56](https://github.com/fredvs/freepascal-ootb/assets/804555/239fe649-3ccc-4fa5-ac17-bd54b3831a09)

Now, if I go in one of my project folders and launch fp from there, it asks me to create a fp.cfg and then it asks me if I want to copy the existing. I agree and it copies the fp.cfg from its main folder, the folder where the executable resides. That cfg file contains the correct folders with $FPCBINDIR...

**Comment by fredvs** [comment]
Hum, no, surely I miss something because if fp is inside /tools/ and use $FPCBINDIR the compilation fails.
It works only if fp has a subdirectory /units ( so if placed alongside with fpc-ootb where there is subdirectory /units).

But maybe I miss something and if you are happy with fp inside /tools it is better (but I have serious doubt about the creation of new fp.cfg if fp was loaded from a directory where there is no fp.cfg).

**Comment by funlw65** [comment]
As you see, I don't have it inside tools. I moved it into its own folder, as a standalone install.

**Comment by funlw65** [comment]
where it has its subfolders as units, libs, inc, etc.... I wouldn't have it in any other way. My interest is to have it functional. It is separated of the fpc-ootb...

**Comment by fredvs** [comment]
> Ok to put you into context, this is the tree of my fp installation:

Ha, ok, you have copy the /units folder inside the directory of fp.
Yes, of course it will work (but when a new fp.cfg is created, there should no have $FPCBINDIR.)

I was thinking about something more out-of-the-box, where you dont have to tweak anything, just load fpide from the directory of the release.

**Comment by funlw65** [comment]
I did not created a new, I said it to copy the existing because, again, I am interested in a functional package.

**Comment by fredvs** [comment]
Conclusion: I miss something.

**Comment by funlw65** [comment]
To address the problem of your package, definitely fp must go in the same folder as fpc-ootb and not in a tools folder... is not a tool, is the ide itself. It is at the same rank as the fpc. If you want it running out of the box.

**Comment by fredvs** [comment]
Here the result of the new fp.cfg if I dont use the last patch: it uses the directory of official fpc.
I am lost, maybe tonight I will have clearer mind and understand why it is working for you.

![fpide_cfg3](https://github.com/fredvs/freepascal-ootb/assets/3421249/5275331b-12c3-41eb-ad00-df7f4702b57b)

**Comment by funlw65** [comment]
Do you need me to archive and post my package?

**Comment by fredvs** [comment]
>Do you need me to archive and post my package?

Ho, yes it will be great, maybe I will understand your trick.

**Comment by funlw65** [comment]
https://drive.google.com/drive/folders/1O5kkWh1YKSV3pICJdf9FwylS64cKrIDB?usp=drive_link

**Comment by fredvs** [comment]
By the way, when I announced to fpc mailing list if it is ok if I do a fork of fpc with all the modifs (that they dont want to commit to official fpc), Sven (a compiler dev) said: "OK, but please use a different name for fpc so if people ask question they will ask about your version of fpc".

It is the reason why I renamed fpc into fpc-ootb.

And I should do the same for FP IDE because it uses the patched source of fpc-ootb.
So I was thinking to rename it into fpide-ootb.

**Comment by fredvs** [comment]
> https://drive.google.com/drive/folders/1O5kkWh1YKSV3pICJdf9FwylS64cKrIDB?usp=drive_link

Many thanks for this.
When I try to download it a message said that I need a confirmation, I clicked ok and a message said that I will receive a mail. I suppose that you need to sent me a confirmation.

Anyway, I have to work now and will try your version tonight.

Thanks for your patience.

Fre;D

**Comment by funlw65** [comment]
you have to unzip it in your home folder... preferably.

**Comment by funlw65** [comment]
Sorry, it was restricted. Now is for everyone that has the link
https://drive.google.com/file/d/16fKhW8XAL7pn6mkAl9Jj-amK3RcSencL/view?usp=sharing

Mea culpa.

**Comment by fredvs** [comment]
OK, I can download it now, thanks.

(But it is 3x the size of release fpc-ootb!)

Write you later tonight when I have tested it.

Fre;D

**Comment by funlw65** [comment]
Now I realize, one of the projects is an example from the Ray4Laz package of Guvacode (Gunko Vadim), it has a subfolder with resources, it contains all the resources of all the examples... it was easier for me to copy it there instead of selecting exactly the needed files for that specific project, Only after I zipped it, I realized how big it is.  Well, next time...

Edit: other heavy files are the books about Turbo Pascal 6, 7, and Turbo Vision, in PDF format.

**Comment by fredvs** [comment]
Maybe I get why for your new config files are using the fp.cfg that is in the same directory as fpide.

Is official fpc installed on your system on **/usr/lib/fpc/'+version_string** ?

If no, it explain why because on my system official fpc is installed on  **/usr/lib/fpc/'+version_string**.

And in **fpini.pas,** the search order of directory of default fp.cfg places official fpc version before directory of FP IDE.

```
    SystemIDEDir:=FExpand(DirOf(system.paramstr(0))+'../lib/fpc/'+version_string+'/ide/text');
     If Not ExistsDir(SystemIDEdir) Then
      SystemIDEDir:='/usr/lib/fpc/'+version_string+'/ide/text';
   
  If Not ExistsDir(SystemIDEdir) Then
    begin
    SystemIDEDir:=FExpand(DirOf(system.paramstr(0))+'../lib64/fpc/'+version_string+'/ide/text');
    If Not ExistsDir(SystemIDEdir) Then
      SystemIDEDir:='/usr/lib/fpc/'+version_string+'/ide/text';
    end;

 ...

```

And it was the reason of last patch.
But I prefer your architecture, place the default fp.cfg in the directory of fpide (vs in /tools/ide_text like in my last patch).

Hum, this maybe, I must first check all this tonight.

Write you later.

**Comment by funlw65** [comment]
I don't have a fpc installed anywhere. No system wide installation. That is why I was delighted by your package. It allows installation in home folder, without the need for admin rights. And is easy to work with and install new units. It is how it worked in MS-DOS back in those days.

**Comment by funlw65** [comment]
Plus, this kind of approach is a must when you use fpc (not this package, but one developed for Cortex core) to program microcontrollers - easy of use, easy to hack, easy to add new libraries, easier for when you need to specify PATHS and a lot of files. The simpler the folder tree and shorter the filenames, the better and faster to get the job done.

**Comment by fredvs** [comment]
Ha, ok, ok, ok, you have a virgin system so there is no problem to load default fp.cfg and fp.ini.

Anyway I will fix the last patch to enable the default fp.cfg and fp.ini in the same directory as fpide (not /tools/fpide_text), even if official fpc version is installed.

And rename it into fpide-ootb (to make everybody happy).
[EDIT] Done: https://github.com/fredvs/freepascal-ootb/releases/tag/3.2.2

Many thanks for your light.

Fre;D

**Comment by funlw65** [comment]
Thank you as well!

**Comment by fredvs** [comment]
By the way, if you use msegui, MSEide and ideU can run also without any installation on the system, just download-unzip release in home folder.

And if you dont want that the config file is installed in the user-config folder, just run MSEide / ideU  with exec parameter **"--globstatfile=<path_of_ideU.stat>"**

Example:

`> ideU --globstatfile=/home/you/somewhere/ideU.sta`

(Note that, for ideU, this feature is enabled in last commit not yet in the release, so the binary must be recompiled.
 For MSEide, the feature exists in the release.)
[EDIT] In last release of ideU  **"--globstatfile=<path_of_ideU.stat>"** is enabled.
            https://github.com/fredvs/ideU/releases/tag/2.14.0


But that is a other story ;-)

**Comment by funlw65** [comment]
Another advantage is that I can have two (or many) fp installs, each in its main folder in $HOME. Why I would need more than one?  Different versions of libraries.

By example, I have the [squabdb](https://gitlab.com/funlw65/squabdb-pas) application that use the older version of raylib, 4.2 and raygui 3.2 and now Guvacode released the 5.0 version which breaks the API quite badly. It will take time to update it (if ever).  So, the fact that it has its own compiler and libraries I am assured that it will compile every time is needed. On the newer package, I can advance with the development of the newer apps without breaking anything. The package that you downloaded contains the newer version of raylib, compiled for each version of OpenGL supported (1.1, 2.1, 3.3, 4.3). It also contains the typhoon relational database library, similar to paradox...

**Comment by funlw65** [comment]
I think I prefer more the Makefile route, maybe because I am used to it from C and microcontroller projects... Anyway, these are the steps I do:

- I download your package for linux 64bit and unzip it in my $HOME folder;
- then I rename the folder of your "fpc-ootb..." as fpc because is shorter and easy to work with;
- I add the units and libs I need then I create the project somewhere else and writhe the following Makefile 

```makefile
LIBPATH=-Fl$(HOME)/fpc/libs
INCPATH=-Fi$(HOME)/fpc/inc
UNITSPATH=-Fu$(HOME)/fpc/units/x86_64-linux/*
FPCFLAGS=-TLinux -Mfpc -Sgisc -CgX -Os -CpATHLON64 -OpCOREI -XS -g- -p- -b- -l -vewnhibq 
# your program executable name 
PROGRAM=squabdb

all:
	~/fpc/fpc-ootb-64 $(LIBPATH) $(INCPATH) $(UNITSPATH) $(FPCFLAGS)  $(PWD)/$(PROGRAM).pas

clean:
	rm *.o *.ppu $(PROGRAM)


```

If is about an open source project, apart of packaging it for github, I repackage your fpc with all the units and static libs needed and make it available for the user to be able to recompile the project.  

The Makefile is perfectible, of course, but I think is a good start, This way you can move the folder of your project wherever you want, edit it with Geany IDE or any editor that has Pascal highlighting syntax and then compile with make without problems.

**Comment by funlw65** [comment]
BTW Fred,

I use Typhoon Relational Database Library, which is a C library with a pascal unit. The user has to define the tables, structures and indexes in a specific language which is then processed with a provided tool and a C header with those definitions is created. The Pascal programmer must convert that C header in a Pascal unit using h2pas tool, which is not present in your package, So, it will be nice if you add it in the tools subfolder, thank you!

**Comment by fredvs** [comment]
Where is the source of h2pas?

**Comment by funlw65** [comment]
regarding the sources of ootb, in utils/h2pas

**Comment by fredvs** [comment]
> regarding the sources of ootb, in utils/h2pas

OK, just added in zip assets of https://github.com/fredvs/freepascal-ootb/releases/tag/3.2.2

**Comment by funlw65** [comment]
Thank you very much Fred!
